### PR TITLE
Added support for zstd compression

### PR DESCRIPTION
**Pull request by Miniwoffer**
- Added support for zstd 

- <del>Changed how compression level setting is handled, it is now handled on a algorithm basis. And warns instead off failing when out of range. (This was needed since zstds compression level range is decided at runtime). </del>

### CODE REVIEW

**Comment by cmouse** [comment]
Thanjs. We'll look into this. Please be patient.

File: src/lib-compression/istream-zstd.c [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	size_t to_read = ZSTD_decompressStream(
+		zstream->dstream, &zstream->output, &zstream->input);
+
+	if (ZSTD_isError(to_read)) {
+		return to_read;
+		// i_fatal("ZSTD_decompressStream():
+		// %s",ZSTD_getErrorName(to_read));
+	}
+	zstream->next_read = to_read;
+}
```
Comment (cmouse) [inline src/lib-compression/istream-zstd.c]: missing return.

File: src/lib-compression/istream-zstd.c [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+				}
+				return -1;
+			}
+			if (size == 0) {
+				break;
+			}
+			if (size > zstream->input.pos) {
+				size = zstream->input.pos;
+			}
+			memcpy((void *)((long)zstream->input.src +
```
Comment (cmouse) [inline src/lib-compression/istream-zstd.c]: use PTR_OFFSET

File: src/lib-compression/istream-zstd.c [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+static size_t i_stream_zstd_decompress(struct zstd_istream *zstream)
+{
+	struct istream_private *stream = (struct istream_private *)zstream;
+	const unsigned char *data;
+	size_t size;
+
+	if (zstream->next_read || !stream->istream.eof) {
+		// move everything to back of buffer
+		memmove(zstream->input.src,
+			(void *)((long)zstream->input.src + zstream->input.pos),
```
Comment (cmouse) [inline src/lib-compression/istream-zstd.c]: use CONST_PTR_OFFSET. Also you can use PTR_OFFSET(zstream->input.src, 0) to non-constify it.

File: src/lib-compression/ostream-bzlib.c [hunk 4]
```diff
@@ -210,7 +210,10 @@ struct ostream *o_stream_create_bz2(struct ostream *output, int level)
 	struct bzlib_ostream *zstream;
 	int ret;
 
-	i_assert(level >= 1 && level <= 9);
+	if (level < 1 || level > 9) {
```
Comment (cmouse) [inline src/lib-compression/ostream-bzlib.c]: the assert is here for a reason.

File: src/plugins/zlib/zlib-plugin.c [hunk 5]
```diff
@@ -357,9 +357,9 @@ static void zlib_mail_user_created(struct mail_user *user)
 	}
 	name = mail_user_plugin_getenv(user, "zlib_save_level");
 	if (name != NULL) {
-		if (str_to_uint(name, &zuser->save_level) < 0 ||
-		    zuser->save_level < 1 || zuser->save_level > 9) {
-			i_error("zlib_save_level: Level must be between 1..9");
+		if (str_to_uint(name, &zuser->save_level) < 0) {
+			i_error("zlib_save_level: Failed to parse "
```
Comment (cmouse) [inline src/plugins/zlib/zlib-plugin.c]: the previous behaviour is preferred.

**Comment by cmouse** [comment]
zstd has `ZSTD_initCStream(zstream->cstream, level)` which lets you decide the compression level.

**Comment by Miniwoffer** [comment]
Thanks for the review,

File: src/lib-compression/istream-zstd.c [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+static void i_stream_zstd_close(struct iostream_private *stream,
+				bool close_parent)
+{
+	struct zstd_istream *zstream = (struct zstd_istream *)stream;
+
+	ZSTD_freeDStream(zstream->dstream);
+	zstream->dstream = NULL;
+
+	i_free(zstream->input.src);
```
Comment (Miniwoffer) [inline src/lib-compression/istream-zstd.c]: This produces a warning in gcc 8.2.1 

```
istream-zstd.c:34:23: warning: passing argument 2 of ‘p_free_internal’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  i_free(zstream->input.src);
```
Any suggestions on how to resolve it without making a temporary pointer?

**Comment by cmouse** [comment]
Hi!

We have now merged our own implementation. Thank you for your efforts!

**Pull request closed without merging**